src/openvic-simulation/types/Colour.hpp

```diff
@@ -1,42 +1,218 @@
-		std::ostringstream stream;
-		stream << std::uppercase << std::hex << std::setfill('0') << std::setw(!alpha ? 6 : 8) << colour;
-		return stream.str();
-	}
+		constexpr basic_colour_t() {}
+		constexpr basic_colour_t(value_type p_value) : value(p_value) {}
+		constexpr explicit operator value_type() const {
+			return value;
+		}
+
+		static constexpr basic_colour_t from_float(float f, float min = 0.0f, float max = 1.0f) {
+			return static_cast<value_type>(std::clamp(min + f * (max - min), min, max) * 255.0f);
+		}
+
+		static constexpr basic_colour_t from_fraction(int n, int d, float min = 0.0f, float max = 1.0f) {
+			return from_float(static_cast<float>(n) / static_cast<float>(d), min, max);
+		}
+
+		static constexpr basic_colour_t from_float_to_alpha(float a) {
+			return from_float(a) << 24;
+		}
+
+		inline std::string to_hex_string(bool alpha = false) const {
+			using namespace std::string_view_literals;
+			static constexpr std::string_view digits = "0123456789ABCDEF"sv;
+
+			const std::int8_t length = alpha ? 8 : 6;
+			std::array<char, 8> address;
+
+			for (std::size_t i = 0, j = (length - 1) * 4; i < length; ++i, j -= 4) {
+				address[i] = digits[(value >> j) & 0x0f];
+			}
+			return { address.data() };
+		}
+
+		constexpr component_return_type red() const {
+			return (value >> red_shift) & component;
+		}
+
+		constexpr component_return_type green() const {
+			return (value >> green_shift) & component;
+		}
+
+		constexpr component_return_type blue() const {
+			return (value >> blue_shift) & component;
+		}
+
+		constexpr component_return_type alpha() const {
+			return (value >> alpha_shift) & component;
+		}
+
+		constexpr bool is_null() const {
+			return *this == null;
+		}
+
+		constexpr auto operator<=>(const basic_colour_t& rhs) const = default;
+		constexpr auto operator<=>(value_type rhs) const {
+			return *this <=> basic_colour_t { rhs };
+		}
+
+		constexpr basic_colour_t operator<<(value_type rhs) const {
+			return value << rhs;
+		}
+		constexpr basic_colour_t operator>>(value_type rhs) const {
+			return value >> rhs;
+		}
+		constexpr basic_colour_t operator|(value_type rhs) const {
+			return value | rhs;
+		}
+		constexpr basic_colour_t operator&(value_type rhs) const {
+			return value & rhs;
+		}
+
+		constexpr basic_colour_t& operator<<=(value_type rhs) {
+			value <<= rhs;
+			return *this;
+		}
+		constexpr basic_colour_t& operator>>=(value_type rhs) {
+			value >>= rhs;
+			return *this;
+		}
+		constexpr basic_colour_t& operator|=(value_type rhs) {
+			value |= rhs;
+			return *this;
+		}
+		constexpr basic_colour_t& operator&=(value_type rhs) {
+			value &= rhs;
+			return *this;
+		}
+
+		friend constexpr basic_colour_t operator<<(value_type lhs, const basic_colour_t& rhs) {
+			return lhs << rhs.value;
+		}
+		friend constexpr basic_colour_t operator>>(value_type lhs, const basic_colour_t& rhs) {
+			return lhs >> rhs.value;
+		}
```

Comment:
Do we need shift operators with a `basic_colour_t` right hand side? I would consider someone storing a number of bits to shift in a colour variable to be a mistake/discouraged. Same applies to the `<<`/`>>` and `<<=`/`>>=` operators below.